### PR TITLE
HMAC v2 java library

### PR DESCRIPTION
**Pull request by g3p3nk**
-) various bug fixes
-) added X-Authorization-Content-SHA256
-) added X-Server-Authorization-HMAC-SHA256
-) added timestamp check

### CODE REVIEW

File: src/main/java/com/acquia/http/HMACFilter.java [hunk 1]
```diff
@@ -81,6 +129,24 @@ public void destroy() {
 
     }
 
+    /**
+     * Check if timestamp is within tolerance (900 seconds)
+     * 
+     * @param unixTimestamp
+     * @return non-zero if timestamp is outside tolerance (positive if in the future; negative in the past); otherwise return zero
+     */
+    public int compareTimestampWithinTolerance(long unixTimestamp) {
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACFilter.java]: Why is this method public?

File: src/main/java/com/acquia/http/HMACHttpRequestInterceptor.java [hunk 2]
```diff
@@ -85,20 +85,29 @@ public void setCustomHeaders(String[] customHeaders) {
...
+            throw new IOException("Fail to sign request message", e);
         }
 
-        authHeader.setSignature(encryptedMessage);
-        request.setHeader("Authorization", authHeader.toString()); //set it with encrypted signature
+        authHeader.setSignature(signedRequestMessage);
+        request.setHeader(HMACMessageCreator.PARAMETER_AUTHORIZATION, authHeader.toString()); //set it with encrypted signature
+
+        //set context for response interceptor
+        context.setAttribute("httpVerb", request.getRequestLine().getMethod().toUpperCase());
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACHttpRequestInterceptor.java]: These strings should be constants that are used in this class and the http response interceptor class

File: src/main/java/com/acquia/http/HMACHttpResponseInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            if (xAuthorizationTimestamp == null) {
+                throw new HttpException("Error: No xAuthorizationTimestamp in the HTTP context.");
+            }
+
+            //get server response body
+            String responseContent = "";
+            final HttpEntity entity = response.getEntity();
+            if (entity != null && entity.getContentLength() > 0) {
+                //response body can only be consumed once - so copy this somewhere
+                final ByteArrayOutputStream baos = HMACUtil.convertInputStreamIntoByteArrayOutputStream(
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACHttpResponseInterceptor.java]: Why not use entity#writeTo method?

File: src/main/java/com/acquia/http/HMACMessageCreator.java [hunk 4]
```diff
@@ -102,15 +111,15 @@ public String createMessage(HttpServletRequest request) throws IOException {
     }
 
     /**
-     * Create the message based on the given HTTP request to be sent and the list of custom header names.
+     * Create response signature message
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACMessageCreator.java]: Why did you change the javadoc?

File: src/main/java/com/acquia/http/HMACMessageCreator.java [hunk 5]
```diff
@@ -45,13 +50,13 @@ public HMACMessageCreator() {
     }
 
     /**
-     * Create the message based on the given HTTP request received and list of custom headers.
+     * Create response signature message
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACMessageCreator.java]: Why did you change the javadoc?

File: src/main/java/com/acquia/http/HMACMessageCreator.java [hunk 6]
```diff
@@ -182,7 +213,7 @@ protected String createMessage(HttpRequest request, HMACAuthorizationHeader auth
     }
 
     /**
-     * Create the message based on the given components of the request.
+     * Helper method to create request signature message
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACMessageCreator.java]: Why did you change the javadoc?

File: src/main/java/com/acquia/http/HMACMessageCreator.java [hunk 7]
```diff
@@ -231,54 +263,66 @@ private String createMessage(String httpVerb, String host, String path, String q
...
+     * 
+     * @param xAuthorizationContentSha256
+     * @param contentLength
      * @return
      * @throws IOException 
      */
-    private byte[] convertInputStreamIntoBtyeArray(InputStream inputStream) throws IOException {
-        if (inputStream == null) {
-            return null;
+    private boolean isPassingRequestBody(int contentLength, String xAuthorizationContentSha256,
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACMessageCreator.java]: Consider splitting this method into two (isPassingRequestBody and getRequestBodySha356)

File: src/main/java/com/acquia/http/HMACUtil.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        return baos;
+    }
+
+    /**
+     * Check if timestamp is within tolerance (900 seconds)
+     * 
+     * @param unixTimestamp
+     * @return non-zero if timestamp is outside tolerance (positive if in the future; negative in the past); otherwise return zero
+     */
+    public static int compareTimestampWithinTolerance(long unixTimestamp) {
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACUtil.java]: Why is there two versions of the same method?

File: src/main/java/com/acquia/http/HMACFilter.java (hunk 1 above)
Comment (g3p3nk) [inline src/main/java/com/acquia/http/HMACFilter.java]: I'll change this to protected. This is originally calling HMACUtil, but mockito doesn't allow mocking static method, so I duplicate the code in this class.. HMACServlet is still using HMACUtil ones..
Not sure if this is a good way to do it.. Input?

File: src/main/java/com/acquia/http/HMACMessageCreator.java (hunk 4 above)
Comment (g3p3nk) [inline src/main/java/com/acquia/http/HMACMessageCreator.java]: No real reason other than that I have another method to create signable response message from httpResponse..
Custom headers are not quite relevant at this point, as it isn't being passed in as parameter anymore, but httpRequest is still relevant.. I guess I can put that in.

File: src/main/java/com/acquia/http/HMACMessageCreator.java [hunk 9]
```diff
@@ -264,8 +265,12 @@ private String createSignableRequestMessage(String httpVerb, String host, String
 
         //adding more if needed
         if (this.isPassingRequestBody(contentLength, xAuthorizationContentSha256, requestBody)) {
-            result.append("\n").append(contentType.toLowerCase());
-            result.append("\n").append(xAuthorizationContentSha256);
+            if (this.isValidRequestBody(xAuthorizationContentSha256, requestBody)) {
+                result.append("\n").append(contentType.toLowerCase());
+                result.append("\n").append(xAuthorizationContentSha256);
+            } else {
+                //FIXME: throw error?
```
Comment (acquiachrisnagy) [inline src/main/java/com/acquia/http/HMACMessageCreator.java]: Should throw an IOException with the appropriate message

**Comment by acquiachrisnagy** [comment]
Looks good
